rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/security/ResourceAuthenticationProvider.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright Siemens AG, 2018. Part of the SW360 Portal Project.
+ *
+ * SPDX-License-Identifier: EPL-1.0
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.eclipse.sw360.rest.resourceserver.security;
+
+import lombok.RequiredArgsConstructor;
+import org.apache.commons.lang.time.DateUtils;
+import org.apache.log4j.Logger;
+import org.eclipse.sw360.datahandler.common.SW360Utils;
+import org.eclipse.sw360.datahandler.thrift.users.RestApiToken;
+import org.eclipse.sw360.datahandler.thrift.users.User;
+import org.eclipse.sw360.rest.resourceserver.user.Sw360UserService;
+import org.jetbrains.annotations.NotNull;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Profile;
+import org.springframework.security.authentication.AuthenticationProvider;
+import org.springframework.security.authentication.AuthenticationServiceException;
+import org.springframework.security.authentication.CredentialsExpiredException;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.AuthenticationException;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.crypto.bcrypt.BCrypt;
+import org.springframework.security.web.authentication.preauth.PreAuthenticatedAuthenticationToken;
+import org.springframework.stereotype.Component;
+
+import java.util.Date;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static java.lang.Math.min;
+import static org.eclipse.sw360.rest.resourceserver.Sw360ResourceServer.*;
+
+@Profile("!SECURITY_MOCK")
+@Component
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class ResourceAuthenticationProvider implements AuthenticationProvider {
+
+    private static final Logger log = Logger.getLogger(ResourceAuthenticationProvider.class);
+
+    @NotNull
+    private final Sw360UserService userService;
+
+    @Override
+    public Authentication authenticate(Authentication authentication) throws AuthenticationException {
+        if (authentication.isAuthenticated()) {
+            log.trace("Authentication already authenticated");
+            return authentication;
+        }
+
+        // Get the corresponding sw360 user and restApiToken based on entered token
+        String token = (String) authentication.getCredentials();
+        String tokenHash = BCrypt.hashpw(token, API_TOKEN_HASH_SALT);
+        User sw360User = getUserFromTokenHash(tokenHash);
+        Optional<RestApiToken> restApiToken = getApiTokenFromUser(tokenHash, sw360User);
```

Comment:
minor comment: I think the naming of this variable is confusing and clashes with `token`.

Reply:
maybe rename database field to `restApiTokenInformation` (because we store more information in this variable),
or use names like `tokenFromAuthentication`, `tokenHashFromAuthentitcation`?